### PR TITLE
Convenience methods for Models and Recipes

### PR DESCRIPTION
**Pull request by kevinaer**
- Added convenience methods for Model class to obtain similar/optimized models
- Added convenience method for Recipe to obtain Model
- Changed Zoo to use a factory approach for creating Recipe, Model, and RecipeTypes to circumvent circular import

Note: Waiting on [PR60](https://github.com/neuralmagic/sparsezoo/pull/60) to land

### CODE REVIEW

**Review by bfineran — DISMISSED** [review]

File: src/sparsezoo/models/zoo.py [hunk 1]
```diff
@@ -89,6 +89,37 @@ class Zoo:
...
+    def constructor(construct_key: str):
+        """
+        :param construct_key: The constructor key
+        :return: Constructor registered with the key
+        """
+        if construct_key not in Zoo._CONSTRUCTORS:
+            raise RuntimeError(f"No constructure registered for {construct_key}")
+        return Zoo._CONSTRUCTORS[construct_key]
+
+    @staticmethod
```
Comment (markurtz) [inline src/sparsezoo/models/zoo.py:104]: NIT: move private static method to after public static

**Review by markurtz — CHANGES_REQUESTED** [review]
I think this might end up being a lot cleaner if we move the functionality of creating and searching for models and recipes into the respective classes on construction of them as well as static and instance methods on them for searching. Then we can have the zoo just call into the Model and Recipe classes rather than us going back and forth between these classes which can quickly become confusing an hard to debug.

On the topic of debugging, would be good to start adding in debug logs when constructing and calling any main functions

File: src/sparsezoo/models/zoo.py [hunk 2]
```diff
@@ -17,77 +17,53 @@
...
-    if valid_params is not None and any(param not in valid_params for param in params):
-        warnings.warn(
-            f"Invalid query string for stub {stub} valid params include {valid_params},"
-            f" given {list(params.keys())}"
-        )
+    _CONSTRUCTORS = {}
 
-    return stub, params
+    @staticmethod
+    def constructor(construct_key: str):
```
Comment (markurtz) [inline src/sparsezoo/models/zoo.py:40]: @kevinaer do we still need the constructor and register functions now?

Comment (kevinaer) [reply src/sparsezoo/models/zoo.py:40]: Ah good catch, we do not. Will get rid of them

**Review by markurtz — APPROVED** [review]

**Review by bfineran — APPROVED** [review]

**Pull request merged**